### PR TITLE
Support -file-compilation-dir

### PR DESCRIPTION
**Pull request by pofat**
This PR adds a new flag `-file-compilation-dir`, which does the same thing as `-ffile-compilation-dir` in Clang.

`swiftc -g -ffile-compilation-dir=. path/to/foo.swift` gives us identical debug info paths regardless of what location we compiled the file from. It's useful to debug correctly using object files built on different machines in different locations.
There's also a long-existed [TODO comment](https://github.com/apple/swift/blob/43324e81307ca8dc51a452d93f24f905a7779626/lib/Frontend/CompilerInvocation.cpp#L1825).

Resolves [SR-5694](https://bugs.swift.org/browse/SR-5694)

### CODE REVIEW

**Comment by pofat** [comment]
@swift-ci please test

**Comment by kastiglione** [comment]
@swift-ci test

File: test/DebugInfo/debug_compilation_dir.swift
Comment (kastiglione) [inline test/DebugInfo/debug_compilation_dir.swift:2]: is `-wmo` relevant to this test?

File: test/DebugInfo/debug_compilation_dir.swift
Comment (kastiglione) [inline test/DebugInfo/debug_compilation_dir.swift:15]: maybe a test for `.` since that's your intended use?

**Review by kastiglione — COMMENTED** [review]
seems proper to me

File: test/DebugInfo/debug_compilation_dir.swift
Comment (pofat) [reply test/DebugInfo/debug_compilation_dir.swift:2]: No, it is not.

File: test/DebugInfo/debug_compilation_dir.swift
Comment (pofat) [reply test/DebugInfo/debug_compilation_dir.swift:15]: Thanks for the advice. Should add that one as well.

**Comment by keith** [comment]
Clang has quite a few flags here that imply different things, both for mappings like the `$PWD=.` style and ones like this. Right now swiftc has `-debug-prefix-map` and `-coverage-prefix-map`.

IMO it is worth considering introducing the equivalent of [`-ffile-prefix-dir .`](https://clang.llvm.org/docs/ClangCommandLineReference.html#cmdoption-clang-ffile-compilation-dir) instead of the debug info only specific variant. But I'm just thinking that because we could potentially avoid the large number of related flags down the line.

File: lib/Frontend/CompilerInvocation.cpp
Comment (adrian-prantl) [inline lib/Frontend/CompilerInvocation.cpp:1827]: is the explicit `std::string()` constructor actually necessary?

File: test/DebugInfo/debug_compilation_dir.swift
Comment (adrian-prantl) [reply test/DebugInfo/debug_compilation_dir.swift:15]: this test will probably not work on Windows. You can check other tests that uses path separators for the correct regex to use.

File: lib/Driver/ToolChains.cpp
Comment (adrian-prantl) [inline lib/Driver/ToolChains.cpp:338]: Does this code need to be added to the Swift implementation of the driver, too? (I haven't looked at the Swift driver code recently so I'm not sure)

**Review by adrian-prantl — COMMENTED** [review]
Thanks, this is a nice improvement! Some comments inline.

File: lib/Frontend/CompilerInvocation.cpp
Comment (pofat) [reply lib/Frontend/CompilerInvocation.cpp:1827]: Frankly speaking, I am not sure.  It still works if I remove the constructor. Do you recommend to remove it if either way works?

File: lib/Driver/ToolChains.cpp
Comment (pofat) [reply lib/Driver/ToolChains.cpp:338]: Yes, I think so. My plan is to address this in the follow up PR.

File: lib/Frontend/CompilerInvocation.cpp
Comment (adrian-prantl) [reply lib/Frontend/CompilerInvocation.cpp:1827]: Yes, please!

File: lib/Driver/ToolChains.cpp
Comment (adrian-prantl) [reply lib/Driver/ToolChains.cpp:338]: I think it may be preferable to do both at once, to keep the implementations in sync at all times.

Comment (pofat) [reply lib/Driver/ToolChains.cpp:338]: Create another [PR](https://github.com/apple/swift-driver/pull/968) for Swift-implemented driver.

**Comment by compnerd** [comment]
Could you please add a test to ensure that the Windows path separator is handled properly?  (`\` vs `/`)

**Comment by rmaz** [comment]
> Could you please add a test to ensure that the Windows path separator is handled properly? (`\` vs `/`)

The existing `CHECK-ABS` test should handle Windows correctly. Would you like to see more tested beyond this?

**Comment by pofat** [comment]
> Yes, I would. If you note the `CHECK-ABS` string is looking at a Unix style separator. I would like to see a test that explicitly tests for Windows style paths e.g. `S:\SourceCache\swift\stdlib\public\core\Array.swift`.

Do you suggest one test for unix and another one for Windows style paths? eg:
```
// RUN: %target-swiftc_driver -g \
// RUN:   -c -debug-compilation-dir /path/to \
// RUN:   %s -o - -emit-ir | %FileCheck --check-prefix=CHECK-ABS-%target-os %s
// RUN: %empty-directory(%t)

func foo() {}

// CHECK-ABS-ios:       !DIFile(filename: "{{.*}}/debug_compilation_dir.swift", directory: "/path/to")
// CHECK-ABS-macosx:    !DIFile(filename: "{{.*}}/debug_compilation_dir.swift", directory: "/path/to")
// CHECK-ABS-watchos:   !DIFile(filename: "{{.*}}/debug_compilation_dir.swift", directory: "/path/to")
// CHECK-ABS-darwin:    !DIFile(filename: "{{.*}}/debug_compilation_dir.swift", directory: "/path/to")
// CHECK-ABS-objc-linux-gnu:   !DIFile(filename: "{{.*}}/debug_compilation_dir.swift", directory: "/path/to")
// CHECK-ABS-native-linux-gnu: !DIFile(filename: "{{.*}}/debug_compilation_dir.swift", directory: "/path/to")
// CHECK-ABS-objc-windows:     !DIFile(filename: "{{[a-zA-Z]:\\(?:[\\\S|*\S]+[\\\/])}}debug_compilation_dir.swift", directory: "/path/to")
// CHECK-ABS-native-windows:   !DIFile(filename: "{{[a-zA-Z]:\\(?:[\\\S|*\S]+[\\\/])}}debug_compilation_dir.swift", directory: "/path/to")
```

**Comment by pofat** [comment]
hi @compnerd , the test for windows path is [added](https://github.com/apple/swift/pull/40735/files#diff-b72180408e3252dc93bb9f51b1dec5bc20e841712f08134dd6ffe5012ad7ee08). Would you trigger the tests? Thank you.

**Comment by kastiglione** [comment]
> IMO it is worth considering introducing the equivalent of -ffile-prefix-dir . instead of the debug info only specific variant. But I'm just thinking that because we could potentially avoid the large number of related flags down the line.

bumping this. @pofat what do you think?

**Comment by kastiglione** [comment]
@Keith the page linked to for `-ffile-prefix-dir` doesn't mention that flag.

**Comment by rmaz** [comment]
> > IMO it is worth considering introducing the equivalent of `-ffile-prefix-dir .` instead of the debug info only specific variant. But I'm just thinking that because we could potentially avoid the large number of related flags down the line.
> 
> bumping this. @pofat what do you think?

Wouldn't we want both? Seems odd that clang would have multiple flags but Swift wouldn't.

**Comment by keith** [comment]
Sorry it's [`-ffile-compilation-dir=`](https://clang.llvm.org/docs/ClangCommandLineReference.html#cmdoption-clang-ffile-compilation-dir)

Fair question for sure, I think clang's setup has gotten a bit complex related to this stuff, so it wasn't that they have many separate flags intentionally, but more because they can't ever remove any for backwards compatibility but that's just my take.

For reference clang has:

```
-ffile-compilation-dir=
-ffile-prefix-map=
-fcoverage-compilation-dir=
-fcoverage-prefix-map=
-fdebug-compilation-dir=
-fdebug-prefix-map=
-fmacro-prefix-map=
```

Where the `ffile*` flags imply all the others

**Comment by pofat** [comment]
Hi @kastiglione @Keith, how about having both flags? 
Based on the [implantation](https://github.com/llvm/llvm-project/blob/10883766300465870b1ad602d9ae9d107596c456/clang/lib/Driver/ToolChains/Clang.cpp#L634) in clang, `-ffile-compilation-dir` looks like a shorter syntax for both `-fdebug-compilation-dir` and `-fcoverage-compilation-dir` to me. 

Personally, I would like to have this flag done in this PR, then do `-coverage-compilation-dir` and `-ffile-compilation-dir` in the follow-up PRs. 

Not sure what's the best strategy to decide what flags should migrate to Swift. Probably just follow what Clang has, IMO.
What do you guys think?

**Comment by keith** [comment]
I will defer to the swift maintainers in general here, but my thought was just instead of doing a 1:1 copy from clang we could try to reduce the user facing complexity here, unless we can come up with compelling reasons for them to be separate.

**Comment by drodriguez** [comment]
@swift-ci please smoke test

**Comment by pofat** [comment]
hi @kastiglione @drodriguez , could anyone of you trigger the test for me again? 
I've encountered a test error (`test/SILGen/moveonly_builtin.swift`). I believed that #40806 fixed it and I rebased my code. 
Thank you very much.

**Comment by pofat** [comment]
> I will defer to the swift maintainers in general here, but my thought was just instead of doing a 1:1 copy from clang we could try to reduce the user facing complexity here, unless we can come up with compelling reasons for them to be separate.

Swift does not have `-fcoverage-compilation-dir` and `-fcoverage-prefix-map`. To avoid confusion, I would like to stick with the original plan. Use `debug-compilation-dir` for debug information path.

**Comment by keith** [comment]
Swift does have `-coverage-prefix-map <prefix=replacement>`, but not the directory version of that flag. When I added that I didn't try to add the directory for the same reason of not adding a lot of flags unless we felt like it was really worth it

**Comment by drodriguez** [comment]
@swift-ci please smoke test

**Comment by pofat** [comment]
> Swift does have `-coverage-prefix-map` (without the `f`), but not the directory version of that flag. When I added that I didn't try to add the directory for the same reason of not adding a lot of flags unless we felt like it was really worth it

My bad. You are right, swift has `-coverage-prefix-map`.
The goal of this PR is to make debug info independent from the absolute path to the build directory. However, the `-prefix-map` flag still requires the build command line to be dependent on the absolute path .
With this flag, "-debug-compilation-dir ." makes it so that both debug info and the compile command itself are independent.

That's another reason for having this flag.

**Comment by keith** [comment]
Yep I'm totally a +1 on having a flag for this, just wondering if we should try to reduce the number of flags in the future, and just have 1 of these unless we have a good reason to keep them separate

**Comment by pofat** [comment]
Yeah I agreed and it's a great suggestion. Should do that once there's support for `-coverage-compilation-dir`.

**Comment by kastiglione** [comment]
> With this flag, `-debug-compilation-dir .` makes it so that both debug info and the compile command itself are independent.

Unless I am missing something, it seems `-file-compilation-dir .` would achieve this as well, plus simplify the flag interface.

**Comment by keith** [comment]
> plus simplify the flag interface.

Exactly. I think there are 2 clear paths we could take:

1. Merge this as is, then potentially add `-coverage-compilation-dir` and/or `-file-compilation-dir` later
2. Update this PR to add `-file-compilation-dir` now, and potentially add the more specific flags in the future if there is a need for them

I'll leave it at this and let someone from the Swift team make a call

**Comment by pofat** [comment]
> 
> Unless I am missing something, it seems `-file-compilation-dir .` would achieve this as well, plus simplify the flag interface.

This flags also includes `-coverage-compilation-dir` which swift does not support now. Won't it be confused for someone who has used `-ffile-compilation-dir` in clang?

**Comment by kastiglione** [comment]
The options as I see it:
1. Update this PR to support coverage (given that `-coverage-prefix-map` exists, maybe it's not that much work, I haven't looked)
2. Change the name of the flag to `-file-compilation-dir`, but don't add support for coverage yet, leaving it as a TODO/Issue
3. Keep this PR as is, and add `-debug-compilation-dir`

1 is best.

Of 2 and 3, they both have a downside. The downside of 2 would hopefully be short term. Someone would do a followup fix. The downside of 3 (that `-debug-compilation-dir` is unnecessary if `-file-compilation-dir` exists), would be long term. For that reason, I think option 1 and 2 are best. If 2 is chosen, then the flag's docstring could say which types of files it supports and does not yet support. Ex "currently supports debug info, not coverage…".

**Comment by pofat** [comment]
> 1. Update this PR to support coverage (given that `-coverage-prefix-map` exists, maybe it's not that much work, I haven't looked)
> 2. Change the name of the flag to `-file-compilation-dir`, but don't add support for coverage yet, leaving it as a TODO/Issue
> 3. Keep this PR as is, and add `-debug-compilation-dir`
> 

Thanks for the clear explaining of all options! Took option 2 and updated this PR. Also fixed windows test.

**Comment by kastiglione** [comment]
@swift-ci test

**Comment by kastiglione** [comment]
The tests and the PR still mention `-debug-compilation-dir`.

**Comment by pofat** [comment]
@kastiglione I am sorry, might undo some old code while pushing. All fixed.

**Comment by kastiglione** [comment]
@swift-ci test

**Review by adrian-prantl — APPROVED** [review]

**Comment by kastiglione** [comment]
@swift-ci test Windows Platform

**Comment by kastiglione** [comment]
can you update the flag for the swift-driver PR as well, in case that results in a separate set of eyes?

**Comment by pofat** [comment]
> can you update the flag for the swift-driver PR as well, in case that results in a separate set of eyes?

Sure, I can. Working on it.

**Comment by pofat** [comment]
Done updating [`swift-driver`](https://github.com/apple/swift-driver/pull/968)

**Comment by pofat** [comment]
> @swift-ci test Windows Platform

Thanks! I am working on fix the failed windows test case. Will update once I fix it.

**Comment by drodriguez** [comment]
@swift-ci please smoke test

**Comment by pofat** [comment]
@kastiglione failed windows test is fixed. Would you like to review it?

**Review by kastiglione — APPROVED** [review]

**Comment by pofat** [comment]
@adrian-prantl @kastiglione can you help me merger this PR? Thank you!

**Pull request merged**
